utils/cpuload/scheddebug/scheddebug.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheddebug
+
+import (
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/cadvisor/info"
+	"github.com/google/cadvisor/utils"
+)
+
+const (
+	schedDebugPath = "/proc/sched_debug"
+)
+
+var (
+	// Scans cpu number, task group name, and number of running threads.
+	// TODO(rjnagal): cpu number is only used for debug. Remove it later.
+	schedRegExp  = regexp.MustCompile(`cfs_rq\[([0-9]+)\]:(.*)\n(?:.*\n)*?.*nr_running.*: ([0-9]+)`)
+	pollInterval = 1 * time.Second
+)
+
+type SchedReader struct {
+	refreshChan   chan error
+	lastErrorTime time.Time
+	dataLock      sync.RWMutex
+	load          map[string]int
+}
+
+func (self *SchedReader) Start() error {
+	self.refreshChan = make(chan error)
+	self.refresh()
+	go self.housekeep()
+	return nil
+}
+
+func (self *SchedReader) Stop() {
+	self.refreshChan <- nil
+	err := <-self.refreshChan
+	if err != nil {
+		glog.Warning("Failed to stop scheddebug load reader: %s", err)
+	}
+}
+
+// Since load housekeeping and normal container housekeeping runs at the same rate,
+// there is a chance of sometimes picking the last cycle's data. We can solve that by
+// calling this housekeeping from globalhousekeeping if its an issue.
+func (self *SchedReader) housekeep() {
+	ticker := time.Tick(pollInterval)
+	for {
+		select {
+		case <-ticker:
+			self.refresh()
+		case <-self.refreshChan:
+			self.refreshChan <- nil
+			glog.Infof("Exiting housekeeping")
+			return
+		}
+	}
+}
+
+func (self *SchedReader) refresh() {
```

Comment:
WDYT: having this return an "error" and doing the logging in housekeeping where it is called?